src/App/DocumentObject.cpp

```diff
@@ -887,20 +888,93 @@ DocumentObject *DocumentObject::getSubObject(const char *subname,
     return ret;
 }
 
-std::vector<DocumentObject*> DocumentObject::getSubObjectList(const char *subname) const {
+namespace
+{
+std::vector<DocumentObject*>
+getSubObjectListFlatten(const std::vector<App::DocumentObject*>& resNotFlatten,
+                        std::vector<int>* const subsizes,
+                        const App::DocumentObject* sobj,
+                        App::DocumentObject** container,
+                        bool& lastChild)
+{
+    auto res {resNotFlatten};
+    auto linked = sobj->getLinkedObject();
+    if (container) {
+        auto grp = App::GeoFeatureGroupExtension::getGroupOfObject(linked);
+        if (grp != *container) {
+            container = nullptr;
+        }
+        else {
+            if (lastChild && !res.empty()) {
+                res.pop_back();
+                if (subsizes) {
+                    subsizes->pop_back();
+                }
+            }
+            lastChild = true;
+        }
+    }
+    if (linked->getExtensionByType<App::GeoFeatureGroupExtension>(true)) {
+        *container = linked;
+        lastChild = false;
+    }
+    else if (linked != sobj || sobj->hasChildElement()) {
+        // Check for Link or LinkGroup
+        container = nullptr;
+    }
+    else if (auto ext = sobj->getExtensionByType<LinkBaseExtension>(true)) {
+        // check for Link array
+        if (ext->getElementCountValue() != 0) {
+            container = nullptr;
+        }
+    }
+    return res;
+}
+}  // namespace
+
+std::vector<DocumentObject*> DocumentObject::getSubObjectList(const char* subname,
+                                                              std::vector<int>* const subsizes,
+                                                              bool flatten) const
+{
     std::vector<DocumentObject*> res;
     res.push_back(const_cast<DocumentObject*>(this));
-    if(!subname || !subname[0])
+    if (subsizes) {
+        subsizes->push_back(0);
+    }
+    if (!subname || (subname[0] == 0)) {

```

Comment:
Somewhere else in the code, something similar was refactored by expressing it in terms of a comparison to `nullptr`, so `subname == nullptr`. I would actually prefer `!subname` because it is so common, but whatever is chosen, perhaps it could be used consistently in this code. I do prefer the switch from `!subname[0]` to `subname[0] == 0`, but while we're at it, why not making it even more correct by doing `subname[0] == '\0'`, comparing two chars?